### PR TITLE
Allow options and actions to be set for installing packages

### PR DESCRIPTION
**Pull request by MarkGibbons**
OS package installation in recipe nodejs_from_package.rb now
allows installation actions and options to be specified using
attributes.

### Description

Allow options and package resource actions to be specified for os package installs.

Refactored the test structure to eliminate redundant test suites.

Added more complete tests for each suite and created tests for suites that had no tests.

### Issues Resolved
#100 - rewritten for the existing cookbook and slightly different attributes were used.

### Check List

- [ ] All tests pass.
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -41,18 +41,16 @@ jobs:
       matrix:
         os:
           - 'amazonlinux-2'
-          - 'centos-7'
```
Comment (ramereth) [inline .github/workflows/ci.yml:44]: Why are you removing EL7?

File: kitchen.yml [hunk 2]
```diff
@@ -52,13 +53,24 @@ suites:
...
-    run_list: recipe[nodejs]
+  - name: package
+    run_list:
+      - recipe[test::default]
+  - name: source
+    run_list:
+      - recipe[test::source]
+  - name: options
+    includes:
+      - ubuntu-20.04
```
Comment (ramereth) [inline kitchen.yml:64]: Why only this platform?

File: README.md [hunk 3]
```diff
@@ -121,6 +121,8 @@ You can append more specific options to npm command with `attribute :options` ar
 - use an array of options (w/ dash), they will be added to npm call.
 - ex: `['--production','--force']` or `['--force-latest']`
 
+You can specify live_stream true for the resource to have the package install information included in the chef-client log outout for better npm package diagnostics and trouble shooting.
```
Comment (jeffbyrnes) [inline README.md:124]: ♥

File: kitchen.yml (hunk 2 above)
Comment (jeffbyrnes) [reply kitchen.yml:64]: I think b/c `npm` works the same regardless of platform, so we only need to test it once, yes? I.e., the intent behind this test suite is to validate `npm_package`’s new `package_options` property, while `npm_package` working well on all platforms is validated by the `package` suite.

@MarkGibbons, it may make sense to combine the two suites, to reduce the number of suites. I.e., you can get two birds with one stone, by setting:

```yaml
- name: package
  run_list:
    - recipe[test::default]
  attributes:
    nodes:
      packages:
        - nodejs
        - acpid
        - adcli
        package_options:
          acpid: "--force-yes"
        package_action:
          adcli: :nothing
```

Comment (ramereth) [reply kitchen.yml:64]: Sometimes you can run into odd platform specific problems so it's always nice to test them on all just in case you catch something.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (MarkGibbons) [reply .github/workflows/ci.yml:44]: Source build doesn't work on centos7 and I don't have time to fix it right now.

Comment (MarkGibbons) [reply .github/workflows/ci.yml:44]: There's a trade off, test source builds or have centos 7 in the mix.  Is there a way to exclude a particular suite, os combination from the standard tests?

File: kitchen.yml (hunk 2 above)
Comment (MarkGibbons) [reply kitchen.yml:64]: I'm quite willing to test everything.  But as I've added test I've found a lot of stuff that just doesn't work.  I'm using #218 as a ticket to track the things that need to be fixed and added back.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ramereth) [reply .github/workflows/ci.yml:44]: You can exclude per suite.

File: kitchen.yml (hunk 2 above)
Comment (ramereth) [reply kitchen.yml:64]: Are they working only on specific suites and not on others? If so I'd exclude them on the broken ones (including a comment for why). You can see how to do that in the percona cookbook.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (MarkGibbons) [reply .github/workflows/ci.yml:44]: Any idea about the syntax for doing the exclude?  The ci.yaml code generates kitchen test xxxxx for the cross product of the suite and os and blows up if put an exclude in a suite in the kitchen.yml file?

File: kitchen.yml (hunk 2 above)
Comment (MarkGibbons) [reply kitchen.yml:64]: I added centos-7 back to the list of OSs.

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,8 @@
 - Verify the URI of installed packages to help determine if a good URI has been installed
 - Add tests that verify npm-package installed packages
 - Get the example test cookbook working
+- Allow actions and options for OS package installation to be specified as attributes
+- Add the live_stream parameter to the npm_package execution to get better installaction diagnostics
```
Comment (ramereth) [inline CHANGELOG.md:9]: `installaction` is misspelled

Comment (MarkGibbons) [reply CHANGELOG.md]: fixed.

File: kitchen.dokken.yml [hunk 5]
```diff
@@ -4,6 +4,7 @@ driver:
   privileged: true  # because Docker and SystemD
   chef_version: <%= ENV['CHEF_VERSION'] || 'current' %>
   chef_license: accept-no-persist
+  install_strategy: always
```
Comment (ramereth) [inline kitchen.dokken.yml:7]: Any particular reason why this is set?

File: recipes/nodejs_from_package.rb [hunk 6]
```diff
@@ -29,5 +29,8 @@
 end
 
 node['nodejs']['packages'].each do |node_pkg|
-  package node_pkg
+  package node_pkg do
+    action node['nodejs']['package_action'][node_pkg] if node['nodejs']['package_action'][node_pkg]
+    options node['nodejs']['package_options'][node_pkg] if node['nodejs']['package_options'][node_pkg]
+  end
 end
```
Comment (ramereth) [inline recipes/nodejs_from_package.rb:36]: I think this should be rewritten as the following so that package installs happen faster:
```ruby
package node['nodejs']['packages'] do
  action node['nodejs']['package_action'][node_pkg] if node['nodejs']['package_action'][node_pkg]
  options node['nodejs']['package_options'][node_pkg] if node['nodejs']['package_options'][node_pkg]
end
```

File: recipes/nodejs_from_source.rb [hunk 7]
```diff
@@ -24,9 +24,13 @@
 
 case node['platform_family']
 when 'rhel', 'fedora', 'amazon'
-  package %w(openssl-devel python3 tar)
+  package %w(openssl-devel python2 tar)
+  link '/usr/bin/python' do
+    to '/usr/bin/python2'
+    action :create
```
Comment (ramereth) [inline recipes/nodejs_from_source.rb:30]: This is the default action for `link`, so you can go ahead and remove it.

**Review by ramereth — CHANGES_REQUESTED** [review]

File: recipes/nodejs_from_package.rb (hunk 6 above)
Comment (MarkGibbons) [reply recipes/nodejs_from_package.rb:36]: There is a separate set of options for each package. So the suggested code doesn't work.

Comment (ramereth) [reply recipes/nodejs_from_package.rb:36]: Ah you're correct.

File: kitchen.dokken.yml (hunk 5 above)
Comment (MarkGibbons) [reply kitchen.dokken.yml:7]: I was getting invalid chef cert.pem files.  I thought this was doing a reinstall.  It does fix the error but was not intended as a permanent thing.

File: recipes/nodejs_from_source.rb (hunk 7 above)
Comment (MarkGibbons) [reply recipes/nodejs_from_source.rb]: Fixed.

File: kitchen.dokken.yml (hunk 5 above)
Comment (ramereth) [reply kitchen.dokken.yml:7]: I'd recommend it be removed before this gets merged.

File: recipes/nodejs_from_source.rb [hunk 8]
```diff
@@ -24,9 +24,12 @@
 
 case node['platform_family']
 when 'rhel', 'fedora', 'amazon'
-  package %w(openssl-devel python3 tar)
+  package %w(openssl-devel python2 tar)
```
Comment (ramereth) [inline recipes/nodejs_from_source.rb:27]: Why are you preferring python2 over python3?

File: recipes/nodejs_from_source.rb (hunk 7 above)
Comment (ramereth) [inline recipes/nodejs_from_source.rb:30]: The more I look at this, the more I'm concerned about doing this on a system level. This cookbook isn't managing python, it's managing nodejs. Can you elaborate on why this symlink is needed and what it's trying to accomplish?

**Review by ramereth — CHANGES_REQUESTED** [review]

File: recipes/nodejs_from_source.rb (hunk 8 above)
Comment (MarkGibbons) [reply recipes/nodejs_from_source.rb:27]: python3 doesn't work. Ran the test.

Comment (MarkGibbons) [reply recipes/nodejs_from_source.rb:27]: ark and make are using python somewhere deep in the bowels to build from source.  python3 doesn't work.   Just installing python2 or python3 doesn't create a python link.  There are other options I suppose.

File: kitchen.dokken.yml (hunk 5 above)
Comment (MarkGibbons) [reply kitchen.dokken.yml]: Removed it.

File: recipes/nodejs_from_source.rb (hunk 7 above)
Comment (MarkGibbons) [reply recipes/nodejs_from_source.rb:30]: Working on a fix. Ark needs python in a later step.

Comment (ramereth) [reply recipes/nodejs_from_source.rb:30]: Is that a problem with ark? It sounds like it to me.

Comment (MarkGibbons) [reply recipes/nodejs_from_source.rb:30]: ark expects python to be there.  Still working on faking it out.

Comment (ramereth) [reply recipes/nodejs_from_source.rb:30]: I've been meaning to add support to pyenv which would make it easier to install a system python via the package manager. Perhaps that might be helpful and then have ark depend on that?

Comment (MarkGibbons) [reply recipes/nodejs_from_source.rb:30]: The issues are showing up on the centos containers. Centos usually has python2 installed for management (my memory anyway).  For me, source install on a container to make testing work is a really small edge case.  If I don't get it soon I'll block the test and move on, get it later.

Comment (ramereth) [reply recipes/nodejs_from_source.rb:30]: I'm very familiar with CentOS. Let me know if you want my help on figuring it out.

File: recipes/nodejs_from_source.rb (hunk 8 above)
Comment (MarkGibbons) [reply recipes/nodejs_from_source.rb:28]: I changed the install to use a temporary location for the python link.  The source install still installs python2 but there's no permanent link to point to it.  That's seems pretty minimal in terms of system changes.

Comment (ramereth) [reply recipes/nodejs_from_source.rb:28]: Yeah but that still seems like a sub-optimal way to do it. Let me ponder this and see if I can provide some feedback on a better way to do it. I don't like linking stuff in `/tmp` either as that will likely go away on reboots.

**Comment by MarkGibbons** [comment]
I'm hoping the link in /tmp goes away.  We don't need it after the build.

File: recipes/nodejs_from_source.rb (hunk 8 above)
Comment (MarkGibbons) [reply recipes/nodejs_from_source.rb:28]: I don't want to keep the link. It just needs to be there for the build. /tmp or a user specified location seems the most harmless place.  There's very little that is optimal about this cookbook :-). I need to get this one merged to use as a base for issues in the backlog.

File: recipes/nodejs_from_source.rb [hunk 9]
```diff
@@ -24,9 +24,15 @@
...
 case node['platform_family']
 when 'rhel', 'fedora', 'amazon'
-  package %w(openssl-devel python3 tar)
+  directory "node['nodejs']['tmpdir']"
+  package %w(openssl-devel python2 tar)
+  link "#{node['nodejs']['tmpdir']}/python" do
+    to '/usr/bin/python2'
+  end
+  path = ENV['PATH']
+  ENV['PATH'] = "#{path}:#{node['nodejs']['tmpdir']}"
```
Comment (ramereth) [inline recipes/nodejs_from_source.rb:33]: Instead of doing this, I recommend you add the following line below in the `ark` resource:
```
environment(PYTHON: 'python2')
```

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -48,11 +48,13 @@ jobs:
...
-          - 'binary'
-          - 'npm'
+          - 'default'
           - 'npm-embedded'
-          - 'npm-packages'
+          - 'package'
+          - 'source'
+        exclude:
+          - os: 'centos-7'
+            suite: 'source'
```
Comment (ramereth) [inline .github/workflows/ci.yml:57]: It looks like this is due to the fact that the default compiler on EL7 isn't new enough to compile it correctly as noted here: https://github.com/nxhack/openwrt-node-packages/issues/470

A possible workaround would be to use the [devtoolkit package from SCL](https://www.softwarecollections.org/en/scls/rhscl/devtoolset-7/), but that's probably better done in a separate PR using the `yum-centos` cookbook to enable the repository.

I recommend you put a comment here that explains why this is excluded and create an issue to get this properly fixed.

File: recipes/nodejs_from_source.rb [hunk 11]
```diff
@@ -24,9 +24,15 @@
 
 case node['platform_family']
 when 'rhel', 'fedora', 'amazon'
-  package %w(openssl-devel python3 tar)
+  directory "node['nodejs']['tmpdir']"
+  package %w(openssl-devel python2 tar)
```
Comment (ramereth) [inline recipes/nodejs_from_source.rb:28]: I think it might be good to add a comment above this explaining that the installer only works with python2 currently.

Comment (MarkGibbons) [reply recipes/nodejs_from_source.rb:28]: Added

File: .github/workflows/ci.yml (hunk 10 above)
Comment (MarkGibbons) [reply .github/workflows/ci.yml:57]: Issue #218 already exists for tracking adding centos support.

**Comment by MarkGibbons** [comment]
@ramereth @jeffbyrnes as far as I can tell this is ready to merge.

**Review by ramereth — APPROVED** [review]
